### PR TITLE
Issue #7721: Update AbstractChecks to log DetailAST - AnnotationOnSam…

### LINKED ISSUE #7721: Update AbstractChecks to log DetailAST - AnnotationOnSameLine
Child of https://github.com/checkstyle/checkstyle/issues/5777

This issue specifically focuses on AnnotationOnSameLine.

### PR DESCRIPTION
**Pull request by wilcoln**
Resolve #7721: Update AbstractChecks to log DetailAST - AnnotationOnSameLine

Diff report : https://wilcoln.github.io/checkstyle-reports/7721/diff/index.html

I ran regression on all projects and the diff report was so huge (1.2Gb+)  that  `git add` was taking forever to run. Beside, i'm not sure that github will allow me to upload such a load. So I've decided to host detailed reports for only the 8 first projects.

The number of violations reported remains the same.
I assure you that it is the case for all 34 projects, I checked out details for each of them in local.

### CODE REVIEW

**Comment by wilcoln** [comment]
@rnveach I did what you suggested, I commented out the second and third method leaving me with only the first one active. Then I went to the test result file as mentioned in the error message
```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   XpathRegressionAnnotationOnSameLineTest.testOne:83->AbstractXpathTestSupport.runVerifications:210->AbstractItModuleTestSupport.verify:207->AbstractItModuleTestSupport.verify:257 unexpected output: /home/wilcoln/Workspace/Projects/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java:10:5: Annotation 'Annotation' should be on the same line with its target. ==> expected: <0> but was: <3>
[INFO] 
[ERROR] Tests run: 272, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.1:check (sevntu-checkstyle-check) @ checkstyle ---
[INFO] You have 0 Checkstyle violations.
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.5:check (default-check) @ checkstyle ---
[INFO] Loading execution data file /home/wilcoln/Workspace/Projects/checkstyle/target/jacoco.exec
[INFO] Analyzed bundle 'checkstyle' with 538 classes
[INFO] All coverage checks have been met.
[INFO] 
[INFO] --- maven-failsafe-plugin:2.22.2:verify (verify) @ checkstyle ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  06:34 min
[INFO] Finished at: 2020-03-23T00:37:28+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:2.22.2:verify (verify) on project checkstyle: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/wilcoln/Workspace/Projects/checkstyle/target/failsafe-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
```


The file is 
`checkstyle/target/failsafe-reports/org.checkstyle.suppressionxpathfilter.XpathRegressionAnnotationOnSameLineTest.txt`

And its content is 
```
-------------------------------------------------------------------------------
Test set: org.checkstyle.suppressionxpathfilter.XpathRegressionAnnotationOnSameLineTest
-------------------------------------------------------------------------------
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.071 s <<< FAILURE! - in org.checkstyle.suppressionxpathfilter.XpathRegressionAnnotationOnSameLineTest
testOne  Time elapsed: 0.067 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: unexpected output: /home/wilcoln/Workspace/Projects/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java:10:5: Annotation 'Annotation' should be on the same line with its target. ==> expected: <0> but was: <3>
	at org.checkstyle.suppressionxpathfilter.XpathRegressionAnnotationOnSameLineTest.testOne(XpathRegressionAnnotationOnSameLineTest.java:83)

```

**Comment by wilcoln** [comment]
@rnveach ping

**Comment by wilcoln** [comment]
@rnveach, @romani, please can you provide me some pointers?

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineThree.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package org.checkstyle.suppressionxpathfilter.annotationonsameline;
+
+@Ann //warn
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineThree.java]: please use default annotations

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class SuppressionXpathRegressionAnnotationOnSameLineTwo
+{
+    @Ann //warn
+    private List<String> names = new ArrayList<>();
+}
+
+@Target({CONSTRUCTOR, FIELD, METHOD, PARAMETER, TYPE, TYPE_PARAMETER, TYPE_USE}) @interface Ann {
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: no need for custom annotation

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import static java.lang.annotation.ElementType.*;
+
+import java.lang.annotation.Target;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SuppressionXpathRegressionAnnotationOnSameLineTwo
+{
+    @Ann //warn
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: here too

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+{
+    @Annotation int x;
+
+    @Annotation //warn
+    public int getX() {
+        return (int) x;
+    }
+}
+
+@interface Annotation {
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java]: here too

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package org.checkstyle.suppressionxpathfilter.annotationonsameline;
+
+public class SuppressionXpathRegressionAnnotationOnSameLineOne
+{
+    @Annotation int x;
+
+    @Annotation //warn
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java]: here too and 2 lines above

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineThree.java (hunk 1 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineThree.java]: Done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java (hunk 2 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: Done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java (hunk 3 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: Done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java (hunk 4 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java]: Done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java (hunk 5 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java]: Done

Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java]: Currently Generating diff report

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package org.checkstyle.suppressionxpathfilter.annotationonsameline;
+
+public class SuppressionXpathRegressionAnnotationOnSameLineOne
+{
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java]: please move curly brace to previous line to make formatting correct

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package org.checkstyle.suppressionxpathfilter.annotationonsameline;
+
+import static java.lang.annotation.ElementType.*;
+
+import java.lang.annotation.Target;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SuppressionXpathRegressionAnnotationOnSameLineTwo
+{
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: same here

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package org.checkstyle.suppressionxpathfilter.annotationonsameline;
+
+import static java.lang.annotation.ElementType.*;
+
+import java.lang.annotation.Target;
```
Comment (strkkk) [inline src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: 2 import above are not needed

**Comment by wilcoln** [comment]
> In diff report some links to the projects shows 404 for some reason, may bt because of too many violations.

It's because I didn't upload them all. I explained why in PR's first comment.

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java (hunk 8 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: Done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java (hunk 7 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineTwo.java]: Done

File: src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java (hunk 6 above)
Comment (wilcoln) [reply src/it/resources/org/checkstyle/suppressionxpathfilter/annotationonsameline/SuppressionXpathRegressionAnnotationOnSameLineOne.java]: Done

**Comment by strkkk** [comment]
But in this case it makes sense to edit diff report to exclude them (or run diff report only for several projects where not so many violations)

**Comment by wilcoln** [comment]
That would be nice indeed, but is there a way to know how many violations of certain rule a project has?

Plus I like to check report for every single project in local, even if I can't upload for some them later.

**Comment by strkkk** [comment]
> That would be nice indeed, but is there a way to know how many violations of certain rule a project has?

There is no rule, you can only guess. Some checks "contradict" common formatting rules, like this one or RequireThis, so they can cause many violations.

**Comment by wilcoln** [comment]
> There is no rule, you can only guess. Some checks "contradict" common formatting rules, like this one or RequireThis, so they can cause many violations.

Got it, thanks.

**Pull request merged**